File: frontend/src/api/aiInsights.ts
import { apiUrl, humanizeFetchError } from "./client";

const BASE = "/api/v1/worldcup2026/ai-insights";

export type AiInsightPayload = {
  lines_ko: string[];
  disclaimer_ko: string;
};

async function postJson<T>(path: string, body: object = {}): Promise<T> {
  const r = await fetch(apiUrl(`${BASE}${path}`), {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify(body),
  });
  if (!r.ok) {
    let detail = r.statusText;
    try {
      const j = (await r.json()) as { detail?: unknown };
      if (typeof j.detail === "string") detail = j.detail;
    } catch {
      /* ignore */
    }
    throw new Error(`[HTTP ${r.status}] ${detail}`);
  }
  return r.json() as Promise<T>;
}

export function postAiHomeWelcome(): Promise<AiInsightPayload> {
  return postJson<AiInsightPayload>("/home-welcome");
}

export function postAiAGroupLens(): Promise<AiInsightPayload> {
  return postJson<AiInsightPayload>("/a-group-lens");
}

export function postAiSpotlightTeaser(spot: "czech_republic" | "mexico" | "south_africa"): Promise<AiInsightPayload> {
  return postJson<AiInsightPayload>("/spotlight-teaser", { spot });
}

export function postAiHistoryYear(payload: {
  year: number;
  host?: string;
  result_label?: string;
  highlights?: string;
}): Promise<AiInsightPayload> {
  return postJson<AiInsightPayload>("/history-year", payload);
}

export function postAiWc2026Snack(): Promise<AiInsightPayload> {
  return postJson<AiInsightPayload>("/wc2026-snack");
}

export function postAiKoreaNtStory(): Promise<AiInsightPayload> {
  return postJson<AiInsightPayload>("/korea-nt-story");
}

export function postAiPlaygroundWarmup(): Promise<AiInsightPayload> {
  return postJson<AiInsightPayload>("/playground-warmup");
}

export function humanizeAiInsightError(e: unknown): string {
  return humanizeFetchError(e);
}

File: backend/app/api/v1/endpoints/ai_insights.py
"""짧은 LLM 인사이트 카드 — 여러 화면 공통."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from app.services import ai_insights as svc

router = APIRouter(prefix="/worldcup2026/ai-insights", tags=["ai-insights"])


def _handle_ai_error(e: Exception) -> HTTPException:
    msg = str(e)
    if "OPENAI_API_KEY" in msg:
        return HTTPException(status_code=501, detail=msg)
    if msg == "OPENAI_HTTP_ERROR":
        return HTTPException(status_code=502, detail="AI 서버와 통신하지 못했습니다.")
    return HTTPException(status_code=502, detail=msg)


@router.post("/home-welcome")
async def insight_home_welcome() -> dict[str, Any]:
    try:
        return await svc.home_welcome()
    except RuntimeError as e:
        raise _handle_ai_error(e) from e


@router.post("/a-group-lens")
async def insight_a_group_lens() -> dict[str, Any]:
    try:
        return await svc.a_group_fan_lens()
    except RuntimeError as e:
        raise _handle_ai_error(e) from e


class SpotlightBody(BaseModel):
    spot: str = Field(..., description="czech_republic | mexico | south_africa")


@router.post("/spotlight-teaser")
async def insight_spotlight_teaser(body: SpotlightBody) -> dict[str, Any]:
    try:
        return await svc.spotlight_teaser(body.spot)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    except RuntimeError as e:
        raise _handle_ai_error(e) from e


class HistoryYearBody(BaseModel):
    year: int
    host: str = ""
    result_label: str = ""
    highlights: str = ""


@router.post("/history-year")
async def insight_history_year(body: HistoryYearBody) -> dict[str, Any]:
    try:
        return await svc.history_year_take(body.year, body.host, body.result_label, body.highlights)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    except RuntimeError as e:
        raise _handle_ai_error(e) from e


@router.post("/wc2026-snack")
async def insight_wc2026_snack() -> dict[str, Any]:
    try:
        return await svc.wc2026_snack()
    except RuntimeError as e:
        raise _handle_ai_error(e) from e


@router.post("/korea-nt-story")
async def insight_korea_nt_story() -> dict[str, Any]:
    try:
        return await svc.korea_nt_story()
    except RuntimeError as e:
        raise _handle_ai_error(e) from e


@router.post("/playground-warmup")
async def insight_playground_warmup() -> dict[str, Any]:
    try:
        return await svc.playground_warmup()
    except RuntimeError as e:
        raise _handle_ai_error(e) from e
